### PR TITLE
handle unknown return codes without dying

### PR DESCRIPTION
**Pull request by busterb**
I think this is what we want. Could you test to make sure it instead returns the actual unknown error code so we can add it to the windows_error gem?

### CODE REVIEW

**Comment by thelightcosine** [comment]
@bcook-r7 The changes to the status_code method make sense, but why are you still removing the .name call?

**Comment by busterb** [comment]
Look more closely - my patch applies to @rwhitcroft's PR, and I believe it _readds_ the .name call.

**Comment by busterb** [comment]
Feel free to cherry-pick that change rather than the amalgamation here. As @rwhitcroft said this was largely to get the conversation started :)

**Comment by busterb** [comment]
I wasn't sure if we wanted to teach windows_error how to generate an 'Unknown' result itself, rather than an empty set when the lookup fails. I suspect in general there are a lot more errors it doesn't know about.

**Comment by thelightcosine** [comment]
@busterb I don't think that's the situation that we're actually running into WindowsError pulls the entire list from Microsoft's complete list of error codes. I suspect the actual issue we're running into is that 'value' itself is nil which says we didn't get a valid packet back at all

**Comment by thelightcosine** [comment]
@busterb I think the changes you made on status_code are still good changes regardless, i just don't think they'll solve the actual problem reported, at least not totally.

**Comment by busterb** [comment]
According to @rwhitcroft's backtrace, we either got an empty list from windows_error, or we got a non- SMB1/SMB2 packet back, in order to have a nil. So you could be right in that there is more to fix. Maybe he can confirm.

**Comment by busterb** [comment]
With the change I made, this should never return a nil now, even if we have a non SMB1/2 packet. And if we get a repro, this should either say error code -1 (meaning invalid packet type), or identify the possibly missing Hex value.

**Comment by rwhitcroft** [comment]
@busterb Different error now:

@dmaloney-r7 Apologies for the junk PR - I'll let you take it from here.

```
msf auxiliary(smb_login) > set rhosts 192.168.2.119
rhosts => 192.168.2.119
msf auxiliary(smb_login) > set smbuser test
smbuser => test
msf auxiliary(smb_login) > set smbpass test
smbpass => test
msf auxiliary(smb_login) > run

[*] 192.168.2.119:445     - SMB - Starting SMB login bruteforce
[-] Auxiliary failed: NoMethodError undefined method `new' for WindowsError::NTStatus:Module
[-] Call stack:
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.3.0/gems/ruby_smb-0.0.12/lib/ruby_smb/generic_packet.rb:49:in `status_code'
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.3.0/gems/ruby_smb-0.0.12/lib/ruby_smb/client/tree_connect.rb:39:in `smb1_tree_from_response'
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.3.0/gems/ruby_smb-0.0.12/lib/ruby_smb/client/tree_connect.rb:27:in `smb1_tree_connect'
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.3.0/gems/ruby_smb-0.0.12/lib/ruby_smb/client.rb:240:in `tree_connect'
[-]   /usr/share/metasploit-framework/lib/metasploit/framework/login_scanner/smb.rb:115:in `rescue in attempt_login'
[-]   /usr/share/metasploit-framework/lib/metasploit/framework/login_scanner/smb.rb:108:in `attempt_login'
[-]   /usr/share/metasploit-framework/lib/metasploit/framework/login_scanner/smb.rb:72:in `attempt_bogus_login'
[-]   /usr/share/metasploit-framework/modules/auxiliary/scanner/smb/smb_login.rb:84:in `run_host'
[-]   /usr/share/metasploit-framework/lib/msf/core/auxiliary/scanner.rb:135:in `block (2 levels) in run'
[-]   /usr/share/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
[*] Auxiliary module execution completed
```

File: lib/ruby_smb/generic_packet.rb [hunk 1]
```diff
@@ -35,15 +35,20 @@ def packet_smb_version
...
+          value = self.smb_header.nt_status.value
         when 'SMB2'
-          status_code = WindowsError::NTStatus.find_by_retval(self.smb2_header.nt_status.value).first
-        else
-          nil
+          value = self.smb2_header.nt_status.value
+      end
+      status_code = WindowsError::NTStatus.find_by_retval(value).first
+      if status_code.nil?
+        status_code = WindowsError::NTStatus.new("0x#{value.to_s(16)}", value,
```
Comment (busterb) [inline lib/ruby_smb/generic_packet.rb:48]: This should have been WindowsError::ErrorCode.new instead, could you pease try that?

**Comment by busterb** [comment]
It should have been WindowsError::ErrorCode.new instead, sorry! I'm still curious what the error code you're getting is.

**Comment by rwhitcroft** [comment]
Same options as before - looks better now. Added a `puts` before `status_code()`'s return:

```
msf auxiliary(smb_login) > run

[*] 192.168.2.119:445     - SMB - Starting SMB login bruteforce
returning: #<WindowsError::ErrorCode:0x00561d96b1d2a8 @name="STATUS_MORE_PROCESSING_REQUIRED", @value=3221225494, @description="{Still Busy} The specified I/O request packet (IRP) cannot be disposed of because the I/O operation is not complete.">
returning: #<WindowsError::ErrorCode:0x00561d9845ad60 @name="STATUS_LOGON_FAILURE", @value=3221225581, @description="The attempted logon is invalid. This is either due to a bad username or authentication information.">
returning: #<WindowsError::ErrorCode:0x00561d965a34f8 @name="0x5b0002", @value=5963778, @description="Unknown 0x5b0002">
returning: #<WindowsError::ErrorCode:0x00561d965a3138 @name="0x5b0002", @value=5963778, @description="Unknown 0x5b0002">
returning: #<WindowsError::ErrorCode:0x00561d96537a28 @name="0x5b0002", @value=5963778, @description="Unknown 0x5b0002">
returning: #<WindowsError::ErrorCode:0x00561d965375a0 @name="0x5b0002", @value=5963778, @description="Unknown 0x5b0002">
[*] 192.168.2.119:445     - This system does not accept authentication with any credentials, proceeding with brute force
returning: #<WindowsError::ErrorCode:0x00561d96b1d2a8 @name="STATUS_MORE_PROCESSING_REQUIRED", @value=3221225494, @description="{Still Busy} The specified I/O request packet (IRP) cannot be disposed of because the I/O operation is not complete.">
returning: #<WindowsError::ErrorCode:0x00561d9845ad60 @name="STATUS_LOGON_FAILURE", @value=3221225581, @description="The attempted logon is invalid. This is either due to a bad username or authentication information.">
returning: #<WindowsError::ErrorCode:0x00561d96299898 @name="0x5b0002", @value=5963778, @description="Unknown 0x5b0002">
returning: #<WindowsError::ErrorCode:0x00561d96298b28 @name="0x5b0002", @value=5963778, @description="Unknown 0x5b0002">
returning: #<WindowsError::ErrorCode:0x00561d96227630 @name="0x5b0002", @value=5963778, @description="Unknown 0x5b0002">
returning: #<WindowsError::ErrorCode:0x00561d96227158 @name="0x5b0002", @value=5963778, @description="Unknown 0x5b0002">
[-] 192.168.2.119:445     - SMB - Failed: '.\test:test',
[!] 192.168.2.119:445     - No active DB -- Credential data will not be saved!
[*] Scanned 1 of 1 hosts (100% complete)
```

**Pull request merged**